spark/src/main/scala-spark-master/shims/TypeWideningShims.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta
+
+import org.apache.spark.sql.catalyst.expressions.Cast
+import org.apache.spark.sql.types._
+
+/**
+ * Type widening only supports a limited set of type changes with Spark 3.5 due to the parquet
+ * readers lacking the corresponding conversions that were added in Spark 4.0.
```

Comment:
Not sure about the mechanics but shouldn't this go in a `scala-spark-4.0` directory instead of master? What happens when the 4.0 is cut/released?